conmon-rs/server/src/console.rs

```diff
@@ -0,0 +1,205 @@
+//! Console socket functionalities.
+
+#![allow(dead_code)] // TODO: remove me when actually used
+
+use anyhow::{bail, format_err, Context, Result};
+use getset::Getters;
+use log::{debug, error, trace};
+use nix::sys::termios::{self, OutputFlags, SetArg};
+use sendfd::RecvWithFd;
+use std::{
+    io::ErrorKind,
+    os::unix::{fs::PermissionsExt, io::RawFd},
+    path::{Path, PathBuf},
+    sync::{Arc, RwLock},
+    thread::{self, JoinHandle},
+};
+use tempfile::Builder;
+use tokio::{
+    fs,
+    io::{AsyncWriteExt, Interest},
+    net::{UnixListener, UnixStream},
+    runtime::Runtime,
+};
+
+#[derive(Debug, Getters)]
+#[getset(get)]
+pub struct Console {
+    path: PathBuf,
+    handle: JoinHandle<Result<()>>,
+    fd: Arc<RwLock<Option<RawFd>>>,
+}
+
+impl Console {
+    /// Setup a new console socket.
+    pub fn new() -> Result<Self> {
+        let path = Self::temp_file_name("conmon-term-", ".sock")?;
+        let fd = Arc::new(RwLock::new(None));
+        let fd_clone = fd.clone();
+        let path_clone = path.clone();
+        let handle = thread::spawn(move || Self::listen(&path_clone, fd_clone));
+        Ok(Self { path, handle, fd })
+    }
+
+    /// Returns true if the console socket is successfully connected.
+    pub fn is_connected(&self) -> bool {
+        self.fd().read().map(|fd| fd.is_some()).unwrap_or_else(|e| {
+            error!("Unable to retrieve fd lock: {}", e);
+            false
+        })
+    }
+
+    /// Create window resize control FIFO.
+    pub fn setup_fifo() -> Result<()> {
+        // TODO: implement me
+        unimplemented!();
+    }
+
+    /// Generate a the temp file name without creating the file.
+    fn temp_file_name(prefix: &str, suffix: &str) -> Result<PathBuf> {
+        let file = Builder::new()
+            .prefix(prefix)
+            .suffix(suffix)
+            .rand_bytes(7)
+            .tempfile()
+            .context("create tempfile")?;
+        let path: PathBuf = file.path().into();
+        drop(file);
+        Ok(path)
+    }
+
+    fn listen(path: &Path, fd_state: Arc<RwLock<Option<RawFd>>>) -> Result<()> {
+        Runtime::new()?.block_on(async move {
+            let listener = UnixListener::bind(&path)?;
+            debug!("Listening console socket on {}", &path.display());
+
+            // Update the permissions
+            let mut perms = fs::metadata(&path).await?.permissions();
+            perms.set_mode(0o700);
+            fs::set_permissions(&path, perms).await?;
+
+            let stream = listener.accept().await?.0;
+            debug!("Got console socket stream: {:?}", stream);
+
+            Self::handle_fd_receive(stream, path, fd_state).await
+        })
+    }
+
+    async fn handle_fd_receive(
+        mut stream: UnixStream,
+        path: &Path,
+        fd_state: Arc<RwLock<Option<RawFd>>>,
+    ) -> Result<()> {
+        loop {
+            if stream.ready(Interest::READABLE).await?.is_readable() {
```

Comment:
what if we made this `if !stream.ready(Interest::READABLE).await?.is_readable() { continue; }` and then we could  move the rest of the block in?

Reply:
Sound good, changed as suggested :+1: